File: companion/ios/Sources/CameraCapabilities.swift
import AVFoundation

struct RecordingProfile {
    let width: Int32
    let height: Int32
    let fps: Int
    var label: String { "\(height)p@\(fps)" }
}

struct CameraCaps {
    /// Deduped (resolution -> max fps) for normal recording.
    let normalModes: [(w: Int32, h: Int32, maxFps: Double)]
    /// High-frame-rate / slow-motion modes, reported SEPARATELY — never used for normal record.
    let slowMoModes: [(w: Int32, h: Int32, maxFps: Double)]
    let chosen: RecordingProfile
    let chosenFormat: AVCaptureDevice.Format
    let hdrSupported: Bool

    func describe() -> String {
        func fmt(_ list: [(w: Int32, h: Int32, maxFps: Double)]) -> String {
            list.isEmpty ? "none" :
                list.sorted { Int($0.w) * Int($0.h) > Int($1.w) * Int($1.h) }
                    .map { "\($0.w)x\($0.h)@\(Int($0.maxFps))" }
                    .joined(separator: ", ")
        }
        return """
        Chosen normal-record profile: \(chosen.label) (\(chosen.width)x\(chosen.height))
        Normal modes (res@maxfps): \(fmt(normalModes))
        Slow-motion / high-speed (NOT used for normal record): \(fmt(slowMoModes))
        HDR supported on chosen format: \(hdrSupported ? "yes" : "no")
        """
    }
}

enum CameraCapabilities {

    /// Spec preference chain: 4K/60 -> 1080p/60 -> 1080p/30 -> largest available (fps capped 30).
    static func query(for device: AVCaptureDevice) -> CameraCaps? {
        // (w, h, maxFps, format)
        var all: [(Int32, Int32, Double, AVCaptureDevice.Format)] = []
        for f in device.formats {
            let d = CMVideoFormatDescriptionGetDimensions(f.formatDescription)
            let maxFps = f.videoSupportedFrameRateRanges.map { $0.maxFrameRate }.max() ?? 0
            all.append((d.width, d.height, maxFps, f))
        }
        if all.isEmpty { return nil }

        // Pick a format supporting (w,h,fps); among matches prefer the one whose maxFps is
        // smallest-but-sufficient, so a 30fps normal record never grabs a 240fps slow-mo format.
        func find(_ w: Int32, _ h: Int32, _ fps: Double) -> AVCaptureDevice.Format? {
            all.filter { $0.0 == w && $0.1 == h && $0.2 >= fps }
               .sorted { $0.2 < $1.2 }
               .first?.3
        }

        let chosenFormat: AVCaptureDevice.Format
        let chosen: RecordingProfile
        if let f = find(3840, 2160, 60) {
            chosenFormat = f; chosen = RecordingProfile(width: 3840, height: 2160, fps: 60)
        } else if let f = find(1920, 1080, 60) {
            chosenFormat = f; chosen = RecordingProfile(width: 1920, height: 1080, fps: 60)
        } else if let f = find(1920, 1080, 30) {
            chosenFormat = f; chosen = RecordingProfile(width: 1920, height: 1080, fps: 30)
        } else {
            let best = all.max { (Int($0.0) * Int($0.1)) < (Int($1.0) * Int($1.1)) }!
            chosenFormat = best.3
            chosen = RecordingProfile(width: best.0, height: best.1, fps: Int(min(best.2, 30)))
        }

        // Dedup by resolution for the report; split slow-mo (>= 120 fps).
        func dedup(_ src: [(Int32, Int32, Double, AVCaptureDevice.Format)])
            -> [(w: Int32, h: Int32, maxFps: Double)] {
            var byRes: [String: (Int32, Int32, Double)] = [:]
            for m in src {
                let k = "\(m.0)x\(m.1)"
                if let e = byRes[k] { if m.2 > e.2 { byRes[k] = (m.0, m.1, m.2) } }
                else { byRes[k] = (m.0, m.1, m.2) }
            }
            return byRes.values.map { (w: $0.0, h: $0.1, maxFps: $0.2) }
        }

        let normal = dedup(all.filter { $0.2 < 120 })
        let slow = dedup(all.filter { $0.2 >= 120 })

        return CameraCaps(
            normalModes: normal,
            slowMoModes: slow,
            chosen: chosen,
            chosenFormat: chosenFormat,
            hdrSupported: chosenFormat.isVideoHDRSupported
        )
    }
}

File: companion/android/app/src/main/java/com/rxs/recorder/CameraCapabilities.kt
package com.rxs.recorder

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.media.MediaRecorder
import android.os.Build
import android.util.Range
import android.util.Size

/**
 * Runtime camera capability query via Camera2.
 *
 * Phase-2 goal: NEVER claim a mode the device cannot produce. We read the
 * authoritative per-size max frame rate from the stream-configuration map's
 * minimum frame duration, plus the AE target-FPS ranges, and choose a normal
 * recording profile honestly. Slow-motion (high-speed) modes are reported
 * SEPARATELY and never used for normal recording (per spec + HONOR X7d note).
 */

data class RecordingProfile(val width: Int, val height: Int, val fps: Int) {
    val label: String get() = "${height}p@$fps"
}

data class SlowMoProfile(val size: Size, val fpsRanges: List<Range<Int>>)

data class CameraCaps(
    val cameraId: String,
    val normalSizes: List<Size>,
    val maxFpsBySize: Map<Size, Int>,
    val aeFpsRanges: List<Range<Int>>,
    val chosen: RecordingProfile,
    val slowMo: List<SlowMoProfile>,
    val hdr10: Boolean,
    val stabilizationModes: List<Int>
) {
    /** Human-readable dump for the Phase 3–5 device test report. */
    fun describe(): String {
        val sizesStr = normalSizes
            .sortedByDescending { it.width.toLong() * it.height }
            .joinToString(", ") { "${it.width}x${it.height}@${maxFpsBySize[it] ?: 30}" }
        val ae = aeFpsRanges.joinToString(", ") { "[${it.lower},${it.upper}]" }
        val slow = if (slowMo.isEmpty()) "none" else slowMo.joinToString(", ") { sp ->
            "${sp.size.width}x${sp.size.height}@" + sp.fpsRanges.joinToString("/") { "${it.upper}" }
        }
        return buildString {
            appendLine("Back camera id: $cameraId")
            appendLine("Chosen normal-record profile: ${chosen.label} (${chosen.width}x${chosen.height})")
            appendLine("Supported video sizes (size@maxfps): $sizesStr")
            appendLine("AE target FPS ranges: $ae")
            appendLine("HDR10 (10-bit) reported: ${if (hdr10) "yes" else "no"}")
            appendLine("Video stabilization modes: $stabilizationModes")
            appendLine("Slow-motion / high-speed (NOT used for normal recording): $slow")
        }
    }
}

object CameraCapabilities {

    fun query(context: Context): CameraCaps? {
        val cm = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
        val backId = cm.cameraIdList.firstOrNull {
            cm.getCameraCharacteristics(it)
                .get(CameraCharacteristics.LENS_FACING) == CameraCharacteristics.LENS_FACING_BACK
        } ?: cm.cameraIdList.firstOrNull() ?: return null

        val ch = cm.getCameraCharacteristics(backId)
        val map = ch.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP) ?: return null

        val sizes = (map.getOutputSizes(MediaRecorder::class.java) ?: emptyArray()).toList()
        val aeRanges = (ch.get(CameraCharacteristics.CONTROL_AE_AVAILABLE_TARGET_FPS_RANGES)
            ?: emptyArray()).toList()

        // Authoritative per-size max fps from min frame duration (ns).
        val maxFpsBySize = sizes.associateWith { s ->
            val minDur = runCatching { map.getOutputMinFrameDuration(MediaRecorder::class.java, s) }
                .getOrDefault(0L)
            if (minDur > 0L) Math.round(1_000_000_000.0 / minDur).toInt() else 30
        }
        val deviceSupports60 = aeRanges.any { it.upper >= 60 }

        val slow = (map.highSpeedVideoSizes ?: emptyArray()).map { s ->
            SlowMoProfile(s, (map.getHighSpeedVideoFpsRangesFor(s) ?: emptyArray()).toList())
        }

        val stab = (ch.get(CameraCharacteristics.CONTROL_AVAILABLE_VIDEO_STABILIZATION_MODES)
            ?: intArrayOf()).toList()

        val hdr10 = if (Build.VERSION.SDK_INT >= 33) {
            (ch.get(CameraCharacteristics.REQUEST_AVAILABLE_CAPABILITIES) ?: intArrayOf())
                .contains(CameraCharacteristics.REQUEST_AVAILABLE_CAPABILITIES_DYNAMIC_RANGE_TEN_BIT)
        } else false

        return CameraCaps(
            cameraId = backId,
            normalSizes = sizes,
            maxFpsBySize = maxFpsBySize,
            aeFpsRanges = aeRanges,
            chosen = pickProfile(sizes, maxFpsBySize, deviceSupports60),
            slowMo = slow,
            hdr10 = hdr10,
            stabilizationModes = stab
        )
    }

    /** Spec preference chain: 4K/60 → 1080p/60 → 1080p/30 → (else) largest@its max (capped 30). */
    private fun pickProfile(
        sizes: List<Size>,
        maxFps: Map<Size, Int>,
        deviceSupports60: Boolean
    ): RecordingProfile {
        fun supports(w: Int, h: Int, fps: Int): Boolean {
            val s = sizes.firstOrNull { it.width == w && it.height == h } ?: return false
            val cap = maxFps[s] ?: 30
            return cap >= fps && (fps < 60 || deviceSupports60)
        }
        return when {
            supports(3840, 2160, 60) -> RecordingProfile(3840, 2160, 60)
            supports(1920, 1080, 60) -> RecordingProfile(1920, 1080, 60)
            supports(1920, 1080, 30) -> RecordingProfile(1920, 1080, 30)
            else -> {
                val s = sizes.maxByOrNull { it.width.toLong() * it.height } ?: Size(1280, 720)
                RecordingProfile(s.width, s.height, minOf(maxFps[s] ?: 30, 30))
            }
        }
    }
}
